### DISCUSSION TITLE
cli and ncurses

### DISCUSSION THREAD

**Discussion by tappara** [daniele77/cli#215] [General]
Ciao Daniele,
I was wondering if there is a way to make coexist `ncurses` and `cli`.
My idea is to build a console application (TUI) that prints some log in a separate panel (or "window", in `ncurses` wording) from where the command are issued. The need is obvious: the side effect of a command might be verbose and it makes sense to print it elsewhere.
(`ncurses` is one possible choice, and despite being written in C, C++ wrappers are available).

You designed `cli` in a flexible enough way that anyone willing to use a ostream object other than `std::cout` can do that, so that should not be an issue to wrap ncurses into a new ostream object and use that in `cli`

I am instead a little concerned about the stdin stream as, at least in `CliLocalTerminalSession`, you manage the stdin (keyboard) natively, and that should NOT work with `ncurses` (as far as I understand `ncurses`....).

Questions: 

1. if I use `CliFileSession` instead (`CliFileSession` does requires a istream object in the constructor) and I grab the stdin stream from curses and pass it to the `cli` object, in you opinion, will that work ? 
2. Alternatively, do you think it would be possible to modify CliLocalTerminalSession to get the stdin input from a istream object passed in the constructor ?

Thanks for your answer. 
Nicola.

PS: if anybody has a better idea on how to use `cli` for a TUI application with two panels, I'm all ears :-)

**Comment by daniele77** [comment]
Hi Nicola,

If I understand correctly, you want to have the input and output of the CLI on different "windows." In one window, you want to enter the commands, and in the other window, you want to see the output of the commands you entered in the first window.

I'm not very familiar with ncurses, but if it provides a couple of `std::istream` and `std::ostream`, you can use the `CliFileSession` since you can pass them to its constructor.

An alternative approach could be to implement your own `NCursesSession` class by deriving it from `CliSession`. This might be useful if ncurses requires specific commands to print the output in its own window, for example.

However, please keep in mind that `CliSession` (the base class for every session) prints the prompt on the same `std::ostream` where you print the output of the CLI commands. In your case, you might end up having the prompt in the separate output window, whereas I assume you want it in the input window. To address this issue, you can override the `CliSession::Prompt()` method in the `NCursesSession` class so that it uses the input window to display the prompt. If you need assistance with that, please let me know, and I can mark `CliSession::Prompt()` as a virtual function.

**Reply by tappara** [reply]
Hi Daniele,
 
Thanks for your answer. 
I think that using `CliFileSession` would be the most straightforward solution.
If I came up with something generic enough to be reused, I'll share my findings.

Nicola
